security/pkg/pki/ca/ca.go

```diff
@@ -317,9 +329,244 @@ func (ca *IstioCA) SignWithCertChain(csrPEM []byte, subjectIDs []string, ttl tim
+	defer cancel()
+
+	caSecret, scrtErr := config.Client.Secrets(config.CaStorageNamespace).Get(CASecret, metav1.GetOptions{})
+	if scrtErr != nil && !errors.IsNotFound(scrtErr) {
+		log.Errorf("Self-signed Citadel failed to read secret that holds CA certificate: %s. "+
+			"Wait until secret %s:%s can be loaded", scrtErr.Error(), config.CaStorageNamespace, CASecret)
+		ticker := time.NewTicker(config.RetryInterval)
+		for scrtErr != nil {
+			select {
+			case <-ticker.C:
+				if caSecret, scrtErr = config.Client.Secrets(config.CaStorageNamespace).Get(CASecret, metav1.GetOptions{}); scrtErr == nil {
+					break
+				}
+			case <-ctx.Done():
+				log.Errorf("Self-signed Citadel failed to load CA secret %s:%s until timeout.",
+					config.CaStorageNamespace, CASecret)
+				break
+			}
+		}
+	}
+	return caSecret, scrtErr
+}
+
+func (ca *IstioCA) checkAndRotateRootCertForReadOnlyCitadel(
+	config *SelfSignedCARootCertRotationConfig, caSecret *v1.Secret, scrtErr error) RootUpgradeStatus {
+	if scrtErr != nil {
+		log.Errorf("Citadel runs in self-signed root cert read only mode but"+
+			" fail to load secret %s:%s (error: %s), skip cert rotation job",
+			config.CaStorageNamespace, CASecret, scrtErr.Error())
+		return UpgradeSkip
+	}
+
+	if caSecret == nil {
+		log.Info("Root cert does not exist, skip root cert rotate for " +
+			"self-signed root cert read-only Citadel.")
+		return UpgradeSkip
+	}
+
+	rootCertificate := ca.GetCAKeyCertBundle().GetRootCertPem()
+	if !bytes.Equal(rootCertificate, caSecret.Data[RootCertID]) {
+		// If the CA secret holds a different root cert than the root cert stored in
+		// KeyCertBundle, this indicates that the local stored root cert is not
+		// up-to-date. Update root cert and key in KeyCertBundle and config map.
+		log.Infof("Load signing key and cert from existing secret %s:%s", caSecret.Namespace, caSecret.Name)
+		rootCerts, err := appendRootCerts(caSecret.Data[caCertID], config.RootCertFile)
+		if err != nil {
+			log.Errorf("failed to append root certificates (%v)", err)
+			return UpgradeFailure
+		}
+		keyCertBundle, err := util.NewVerifiedKeyCertBundleFromPem(caSecret.Data[caCertID],
+			caSecret.Data[caPrivateKeyID], nil, rootCerts)
+		if err != nil {
+			log.Errorf("failed to create CA KeyCertBundle (%v)", err)
+			return UpgradeFailure
+		}
+
+		if err = updateCACertInConfigmapWithRetry(config, keyCertBundle.GetRootCertPem()); err != nil {
+			log.Errorf("Failed to write self-signed Citadel's root cert "+
+				"to configmap (%s). Node agents will not be able to connect.",
+				err.Error())
+		}
+		ca.setCAKeyCertBundle(keyCertBundle)
+		log.Infof("Updated CA KeyCertBundle using existing public key: %v", string(rootCerts))
+		return UpgradeSuccess
+	}
+	return UpgradeSkip
+}
+
+// shouldRefreshCACert checks CA cert for Self-signed Citadel and informs
+// Citadel to refresh CA cert.
+func (ca *IstioCA) shouldRefreshCACert(config *SelfSignedCARootCertRotationConfig,
```

Comment:
This should be a general util function. Can we write a general one and have here and checking the workload cert both calling it?

Reply:
There is a GetWaitTime() in pkg/util/certutil.go. I changed to use that method here and workloadsecret.go